src/wmtk/io/EdgeMeshReader.cpp

```diff
@@ -0,0 +1,164 @@
+        vertices_parameter(i, 2) = vec_vertices_parameter[i][2];
+    }
+}
+
+void EdgeMeshReader::read_obj(
+    std::vector<std::pair<long, long>>& edges,
+    std::vector<std::vector<double>>& vertices,
+    std::vector<double>& vertices_w,
+    std::vector<std::vector<double>>& vertices_texture,
+    std::vector<std::vector<double>>& vertices_normal,
+    std::vector<std::vector<double>>& vertices_parameter)
+{
+    std::ifstream f(m_filename);
+    if (!f.is_open()) {
+        throw std::runtime_error("can't open file!");
+    }
+
+    std::string buffer;
+    while (!f.eof()) {
+        getline(f, buffer);
+        std::vector<std::string> tokens;
+        std::stringstream line(buffer);
+        std::string token;
+        getline(line, token, ' ');
+        switch (str_to_type(token)) {
+        case V: {
+            std::vector<double> positions;
+            while (getline(line, token, ' ')) {
+                // spdlog::info("lineinfo: {}", token);
+                positions.push_back(std::atof(token.c_str()));
+            }
+            assert(positions.size() >= 3);
+            if (positions.size() == 4) {
+                vertices_w.push_back(positions[3]);
+            }
+            vertices.push_back(positions);
+        }; break;
+        case L: {
+            std::vector<long> segment;
+            while (getline(line, token, ' ')) {
+                segment.push_back(std::atol(token.c_str()));
+            }
+            if (segment.size() > 1) {
+                long v0, v1;
+                v0 = segment[0];
+                for (size_t i = 1; i < segment.size(); ++i) {
+                    v1 = segment[i];
+                    std::pair<long, long> edge;
+                    edge.first = v0;
+                    edge.second = v1;
+                    edges.push_back(edge);
+                    v0 = v1;
+                }
+            }
+        }; break;
+        case VT: {
+            std::vector<double> texture;
+            while (getline(line, token, ' ')) {
+                if (token[0] == '[') {
+                    token.erase(std::remove(token.begin(), token.end(), ']'), token.end());
+                    texture.push_back(std::atof(token.c_str() + 1));
+                } else {
+                    texture.push_back(std::atof(token.c_str()));
+                }
+            }
+            assert(texture.size() >= 2);
+            if (texture.size() == 2) {
+                texture.push_back(0);
+            }
+            vertices_texture.push_back(texture);
+        }; break;
+        case VN: {
+            std::vector<double> normal;
+            while (getline(line, token, ' ')) {
+                normal.push_back(std::atof(token.c_str()));
+            }
+            assert(normal.size() == 3);
+            double standard =
+                sqrt(normal[0] * normal[0] + normal[1] * normal[1] + normal[2] * normal[2]);
+            normal[0] /= standard;

```

Comment:
You can do the normalization using `Eigen::Map`:

```
Eigen::Map<Eigen::Vector3d>(normal).normalize()
```
I am not entirely sure if I got the syntax correct but it should be something like that.

Here is the documentation about `Eigen::Map`:
https://eigen.tuxfamily.org/dox-devel/classEigen_1_1Map.html